pkg/nats/nats.go

```diff
@@ -0,0 +1,180 @@
+		return nil
+	}
+}
+
+func New(c *config.Config, opts ...Option) (*natsSource, error) {
+	n := &natsSource{
+		bufferSize: defaultBufferSize,
+	}
+	for _, o := range opts {
+		if err := o(n); err != nil {
+			return nil, err
+		}
+	}
+
+	n.messages = make(chan *Message, n.bufferSize)
+	n.volumeReader = utils.NewNatsVolumeReader(utils.SecretVolumePath)
+
+	opt := []natslib.Option{
+		natslib.MaxReconnects(-1),
+		natslib.ReconnectWait(3 * time.Second),
+		natslib.DisconnectHandler(func(c *natslib.Conn) {
+			n.logger.Info("NATS disconnected")
+		}),
+		natslib.ReconnectHandler(func(c *natslib.Conn) {
+			n.logger.Info("NATS reconnected")
+		}),
+	}
+
+	if c.TLS != nil {
+		if c, err := utils.GetTLSConfig(c.TLS, n.volumeReader); err != nil {
+			return nil, err
+		} else {
+			opt = append(opt, natslib.Secure(c))
+		}
+	}
+
+	if c.Auth != nil {
+		switch {
+		case c.Auth.Basic != nil && c.Auth.Basic.User != nil && c.Auth.Basic.Password != nil:
+			username, err := n.volumeReader.GetSecretFromVolume(c.Auth.Basic.User)
+			if err != nil {
+				return nil, fmt.Errorf("failed to get basic auth user, %w", err)
+			}
+			password, err := n.volumeReader.GetSecretFromVolume(c.Auth.Basic.Password)
+			if err != nil {
+				return nil, fmt.Errorf("failed to get basic auth password, %w", err)
+			}
+			opt = append(opt, natslib.UserInfo(username, password))
+		case c.Auth.Token != nil:
+			token, err := n.volumeReader.GetSecretFromVolume(c.Auth.Token)
+			if err != nil {
+				return nil, fmt.Errorf("failed to get auth token, %w", err)
+			}
+			opt = append(opt, natslib.Token(token))
+		case c.Auth.NKey != nil:
+			nKeyFile, err := n.volumeReader.GetSecretVolumePath(c.Auth.NKey)
+			if err != nil {
+				return nil, fmt.Errorf("failed to get configured nkey file, %w", err)
+			}
+			o, err := natslib.NkeyOptionFromSeed(nKeyFile)
+			if err != nil {
+				return nil, fmt.Errorf("failed to get NKey, %w", err)
+			}
+			opt = append(opt, o)
+		}
+	}
+
+	n.logger.Info("Connecting to nats service...")
+	if conn, err := natslib.Connect(c.URL, opt...); err != nil {
+		n.logger.Error("Failed to connect to nats server", zap.Error(err))
+		return nil, fmt.Errorf("failed to connect to nats server, %w", err)
+	} else {
+		n.natsConn = conn
+	}
+
+	n.logger.Infof("Subscribing to subject %s with queue %s", c.Subject, c.Queue)
+	if sub, err := n.natsConn.QueueSubscribe(c.Subject, c.Queue, func(msg *natslib.Msg) {
+		readOffset := uuid.New().String()
+		m := &Message{
+			payload:    string(msg.Data),
+			readOffset: readOffset,
+			id:         readOffset,
+		}
+		n.messages <- m
+	}); err != nil {
+		n.logger.Error("Failed to QueueSubscribe nats messages", zap.Error(err))
+		n.natsConn.Close()
+		return nil, fmt.Errorf("failed to QueueSubscribe nats messages, %w", err)
+	} else {
+		n.sub = sub
+	}
+	n.logger.Info("NATS source server started")
+	return n, nil
+}
+
+// Pending returns the number of pending records.
+func (n *natsSource) Pending(_ context.Context) uint64 {
```

Comment:
per our discussion probably just need to change this signature to indicate "pending not available"